[lapack] add metaport lapack to switch lapack impl with an overlay

Neumann-A opened the pull request:
In preparation to add netlibs lapack-reference from #11454

Neumann-A committed:
[lapack-select] add port lapack-select to select the lapack implementation

cenit commented:
in order to be ready even better, you should handle macOS  inside lapack-select port (an empty port) ;)

cenit commented:
in order to be ready even better, do you have any idea to handle osx inside lapack-select?
Whenever we will have lapack-reference, it would be better to use it instead of Accelerate Framework!

Neumann-A committed:
move (!osx) check into lapack select

Neumann-A commented:
@cenit what about all those `openblas (!osx)`? Can they also be removed since lapack will depend on blas?

Neumann-A committed:
correct the platform selection logic so that clapack is not forced.
added external as a feature.
made a dummy project to find the external provided LAPACK

Neumann-A commented:
> in order to be ready even better, do you have any idea on how to handle osx inside lapack-select

@cenit see latest commit. You basically need a select feature with a build-depends which has the platform conditions. The feature is required to not always force the dependency.

cenit commented:
> @cenit what about all those `openblas (!osx)`? Can they also be removed since lapack will depend on blas?

OpenBLAS has improved a lot and we can use it on vcpkg now maybe. For sure, it's not like clapack (which is not acceptable on osx)

cenit commented:
> > in order to be ready even better, do you have any idea on how to handle osx inside lapack-select
> 
> 
> 
> @cenit see latest commit. You basically need a select feature with a build-depends which has the platform conditions. The feature is required to not always force the dependency.

Convoluted but effective. Nice!

Neumann-A commented:
> OpenBLAS has improved a lot and we can use it on vcpkg now maybe. For sure, it's not like clapack (which is not acceptable on osx)

@cenit: The question was more is at a real obenBLAS dependency or is it a indirect dependency due to LAPACK which should not be listed in Build-Depends.

cenit commented:
We don't want blas functions from lapack-reference, so we should disable internal blas and add a real dependency on openblas. Similarly, the cblas feature should disappear from lapack-reference, since the C API for blas will be provided by openblas.
Each one for the best part they provide :)

Neumann-A commented:
> We don't want blas functions from lapack-reference

Yeah that is clear. The question was if the ports actually use BLAS functions or if they only use LAPACK functions and thus only indirectly depend on BLAS.

Neumann-A committed:
fix the oscx regression. Actually need to specify external dependency on OSX

cenit commented:
Ah now I understand what you mean
ports depending on `openblas (!osx), clapack (!osx)` in fact might look suspicious, because depending on `clapack (!osx)` should have been enough, since clapack then internally uses blas, provided for us by openblas. But it wouldn't be so strange if they were using lapack for something, but also blas for other things, independently of their relationship, and thus correctly depending on both anyway.
I would have to check their sources. A quick way would be to scan for find_package(BLAS) and find_package(LAPACK). If only the latter, then the former would be an unnecessary direct dependency

Neumann-A commented on ports/mlpack/CONTROL:
```diff
@@ -1,8 +1,9 @@
 Homepage: https://github.com/mlpack/mlpack
 Description: mlpack is a fast, flexible machine learning library, written in C++, that aims to provide fast, extensible implementations of cutting-edge machine learning algorithms.
-Build-Depends: openblas (!osx), clapack (!osx), boost, armadillo, ensmallen, stb
+Build-Depends: openblas (!osx), lapack-select[core], boost, armadillo, ensmallen, stb
```
Doesn't seem to depend on openblas or lapack directly.

strega-nil commented:
@Neumann-A is there a reason that you're depending on `lapack-select[core]` as opposed to just `lapack-select` in the dependencies?

Neumann-A commented:
@strega-nil: depending on `lapack-select` always means `lapack-select[core,default]` which forces the default feature as a build dependency. So `lapack-select[core,otherfeature]` does not fulfill `lapack-select`

JackBoosY commented:
@strega-nil See #11602, use `core` in `Build-Depends` doesn't work currentlly.

strega-nil commented:
@Neumann-A I see... so we want the `lapack-select` dependency to be fulfilled when the user installs `lapack-select` with her own chosen feature... hrmm...

Neumann-A commented:
@strega-nil vcpkg still chooses a default for the user if it is not explicutly specified. If you ask why something like this is necessary just take a peek in cmakes FindLAPACK.cmake

strega-nil commented:
Right, yeah, I'm just thinking about how best to do this, since at some point I'm hoping that we stop installing default features for dependencies listed with `core`.

ras0219 commented:
Features are designed to be additive, not exclusive; they shouldn't be used to model "exactly-one-of" -- they model "Any subset of". Since this is an "exactly-one-of", it isn't a good fit for features.

Instead, for this situation, our current best approach is to use overlay ports. In this approach, there is an intermediate port exactly like in this PR, ideally named like the interface (perhaps just `lapack` and just `blas`). That port is an empty "alias" for the default implementation:

```
# CONTROL
Source: lapack
Version: 0
Description: Metapackage for packages which provide LAPACK
Build-Depends: clapack (!osx)
```
```
# portfile.cmake
set(VCPKG_POLICY_EMPTY_PACKAGE enabled)
```

Users can then redirect this metapackage by copying the port, replacing the `Build-Depends:` however they like, and using `--overlay-ports=`.

Neumann-A committed:
Merge remote-tracking branch 'upstream/master' into add_lapack_select

Neumann-A committed:
rename port to lapack and use build-depends

strega-nil commented:
cc @ras0219 this look good to you?
